### PR TITLE
[SMD-38] Implement NonUniqueCompositeKeyFailure

### PR DESCRIPTION
**Pull request by Connor-OS**
Add user centered error component for Non-unique composite key failure. This error occurs when users specify duplicate funding sources, project outputs, outcomes or risks.

this commit includes a change to the failure message serialisation to ignore duplicate messages and a refactor of all the failure messages to be based solely on the to_user_cetered_components method.

- [x] Unit tests and other appropriate tests added or updated
- [x] README and other documentation has been updated / added (if needed)
- [x] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### CODE REVIEW

File: core/validation/failures.py [hunk 1]
```diff
@@ -125,15 +127,37 @@ def __str__(self):
...
-    def to_user_centered_components(self) -> tuple[str, str, str, str]:
-        # Funding, Outputs and Outcomes
-        # Funding - Project, Funding Source Name, Funding Source, and
-        # Messages
-        # Funding: You have repeated funding information. You must use a new row for each project, funding source name,
-        # funding type and if its been secured.
-        # Outputs: You must use a new row for each project, funding source name, funding type and if its been secured.
-        # Outcomes
-        return "Unimplemented", "Unimplemented", "Unimplemented", "Unimplemented"
+    def to_user_centered_components(self) -> tuple[str, str, str]:
```
Comment (nsnape) [inline core/validation/failures.py:130]: can we get a docstring please?

File: core/validation/failures.py [hunk 2]
```diff
@@ -390,6 +427,9 @@ def serialise_user_centered_failures(
         # ignore tab and section for pre-transformation failures
         return {"PreTransformationErrors": [message for _, _, message in uc_failures]}
 
+    uc_failures = list(set(uc_failures))
+    uc_failures.sort()
```
Comment (cyrusdobbs) [inline core/validation/failures.py:431]: Could you add a comment to just mention that this is to remove duplicate messages

File: tests/validation_tests/test_failures.py [hunk 3]
```diff
@@ -1,17 +1,19 @@
...
     NonNullableConstraintFailure,
+    NonUniqueCompositeKeyFailure,
     WrongInputFailure,
     WrongTypeFailure,
     serialise_user_centered_failures,
 )
 
 
-def test_invalid_enum_user_centered_failures():
+def test_serialise_user_centered_failures():
```
Comment (cyrusdobbs) [inline tests/validation_tests/test_failures.py:16]: Think we should have 2 tests:
- standard validation failures output is as expected - this should cover different failures from various tabs to ensure the nesting logic is working as expected
- pre-transformation failures output is as expected as this flow is slightly different

**Review by cyrusdobbs — APPROVED** [review]
LGTM 👍

**Pull request merged**